src/services/fightService.js

```diff
@@ -0,0 +1,106 @@
+class FightService {
+  constructor(conn, heroService, dungeonService) {
+    this.conn = conn;
+    this.heroService = heroService;
+    this.dungeonService = dungeonService;
+  }
+
+  async fight(heroId) {
+    if (Number.isNaN(Number(heroId))) return Promise.reject(new Error(412));
+    const hero = await this.heroService.retrieveHeroById(heroId);
+    const dungeonInstance = await this.dungeonService.dungeonInstance(heroId);
+    const dungeon = await this.dungeonService.getDungeonData(dungeonInstance.dungeonId);
+    const currentObstacle = dungeon.obstacles[dungeonInstance.removedObstacles];
+    const healthAtStart = hero.healthact;
+    let failFight = 0;
+    let fight = true;
+    let alive = true;
+    let fightLog = [];
+    let fightResponse = [];
+    let scoutedObstacles = await this.dungeonService.scoutedInstance(heroId);
+    if (hero.healthact === 0) return Promise.reject(new Error(410));
+    if (dungeonInstance.removedObstacles === dungeon.obstacles.length) {
+      return new Promise((resolve) => {
+        resolve(fightResponse = [fightLog = ['No enemy left']]);
+      });
+    }
+    do {
+      if (currentObstacle.healthmax > 0 && hero.healthact > 0) {
+        const damage = Math.max(this.attack(hero.attackmin, hero.attackmax, currentObstacle.defense), 0);
+        currentObstacle.healthmax -= damage;
+        fightLog.push(`${hero.name}(${hero.healthact} HP) hits ${currentObstacle.name} for ${damage}`);
+      } else {
+        fightLog.push(`${hero.name} died!`);
+        fight = false;
+        alive = false;
+        break;
+      }
+      if (hero.healthact > 0 && currentObstacle.healthmax > 0) {
+        const damage = Math.max(this.attack(currentObstacle.attackmin, currentObstacle.attackmax, hero.defense), 0);
+        hero.healthact -= damage;
+        fightLog.push(`${currentObstacle.name}(${currentObstacle.healthmax} HP) hits ${hero.name} for ${damage}`);
+      } else {
+        fight = false;
+        fightLog.push(`${currentObstacle.name} died!`);
+        break;
+      }
+      if (healthAtStart === hero.healthact) {
+        failFight += 1;
+      }
+    }
+
+    while (fight === true && failFight < 20);
+
+    if (alive) {
```

Comment:
by this logic if the enemy doesnt kill you, you kill it no matter what. where do you test the failFight? You use it to exit the infinite loop, but thats it, no?

Reply:
added logic to handle failFight result and modified to check on enemy health too.